### DISCUSSION TITLE
How is the migration to Casual 2.0 going?

### DISCUSSION THREAD

**Discussion by kickingvegas** [kickingvegas/casual#42] [Polls]
For folks who were using the Casual 1.x packages, I'm interested in knowing how well the migration has gone for you. Please provide your feedback, either positive or negative. 

Thanks!

**Comment by kickingvegas** [comment]
If you voted "Not Well", please elaborate why here and hopefully I can help.

**Comment by Stebalien** [comment]
For anyone looking for `use-package` suggestions, just use a single use-package block. There's no need to explicitly use all the sub-packages unless you need key-bindings that map directly to sub-menus. All the top-level menus have autoload cookies.

For example, here's my config (using general for evil keybindings):

```elisp
(use-package casual
  :ensure t
  (:keymaps 'dired-mode-map
   :states 'normal
   "?" 'casual-dired-tmenu)
  (:keymaps 'calc-mode-map
   :states 'normal
   "?" 'casual-calc-tmenu)
  (:keymaps 'ibuffer-mode-map
   :states '(normal motion)
   "?" 'casual-ibuffer-tmenu)
  (:keymaps 'Info-mode-map
   :states 'normal
   "?" 'casual-info-tmenu))

**Reply by kickingvegas** [reply]
@Stebalien thanks for sharing this!

**Comment by kickingvegas** [comment]
Folks - 

I've recently updated the upgrade instructions on the README.org file. It will also be reflected in the description when looking at the `casual` package in `package-list`.

https://github.com/kickingvegas/casual?tab=readme-ov-file#install

**Comment by randomphrase** [comment]
I see lots of `use-package` suggestions, but many seem overly complex to me. YMMV of course! Here's what I settled on for now:

```elisp
(use-package casual
  :after (dired)
  :bind (:map dired-mode-map ("?" . casual-dired-tmenu)
	 :map ibuffer-mode-map ("?" . casual-ibuffer-tmenu) ;; 'h' and '?' both bound to describe-mode, we'll just take the latter
	 :map Info-mode-map ("?" . casual-info-tmenu)
  ))
```

**Comment by deen1** [comment]
I had been using Casual 1.0 &#x2013; specifically, just the `calc` package, which I think was the first one out &#x2013; and hadn't updated for a while. I saw the notice about upgrading to the 2.0 version and the use of `use-package`, so I tried to take care when upgrading.

I removed my previous configuration:

    (use-package casual
      :straight (:type git :host github :repo "kickingvegas/Casual")
      :bind
      (:map calc-mode-map
            ("C-o"   .  'casual-main-menu) )
      )

and backed up the old repository, and then tried the following configuration after restarting:

    (use-package casual
      :straight t
      :bind (:map dired-mode-map ("?" . casual-dired-tmenu)
    	 :map ibuffer-mode-map ("?" . casual-ibuffer-tmenu) ;; 'h' and '?' both bound to describe-mode, we'll just take the latter
    	 :map Info-mode-map ("?" . casual-info-tmenu)
    	 :map calc-mode-map ("C-o"  .  casual-calc-tmenu) )
      )

Unfortunately, while the `dired`, `Ibuffer` and `Info` Casual menus work, I can't get the old `calc` menu to pop up anymore. Instead I get the error

    require: Symbol’s value as variable is void: casual-calc-operators-group

Not sure what to do, I've tried rebuilding the package but it doesn't seem to help.

I'm on `GNU Emacs 29.4 (build 2, x86_64-pc-linux-gnu, GTK+ Version 3.24.43, cairo version 1.18.2) of 2024-12-06`, with `Magit v4.1.3-32-gbfbab898, Transient 0.8.1, Git 2.49.0, Emacs 29.4, gnu/linux`.

**Reply by kickingvegas** [reply]
@deen1 Thanks for sharing this. I appreciate that you’ve taken the time to use Casual. Are you using packages from MELPA stable or ELPA? If so, this is likely the cause.

As of the time of this reply (May 21, 2025), breaking changes have been introduced into the pre-release Transient codebase. The latest Casual release (v2.4.3) takes the pre-release changes (0.9.0) in Transient into account. However in doing so, Casual v2.4.3 breaks for users using any released version of Transient up to 0.8.8.

Some more details on the version skew between Transient and Casual is described in [v2.4.3 · kickingvegas/casual · Discussion #208](https://github.com/kickingvegas/casual/discussions/208).

As far as workarounds you can do today, there are a number of options you can take.

1.  Don’t use Casual until Transient 0.9.0 is released, which should be in the first week of June.
2.  Install the latest snapshot of Transient from MELPA.

Given my current circumstances, I will not be able to issue a fix for MELPA stable in time before Transient 0.9.0 is released. If you can live without Casual, my guidance is to take option 1, as it would least disrupt your configuration.

My apologies for any inconvenience this has brought for you.

**Reply by deen1** [reply]
Thanks for your reply. I see, I have had similar dependency issues with Transient and Magit (which relies heavily on Transient) in the past after building and upgrading Emacs.

I'm happy to wait for the updates to Transient to progress to the main release versions and then try out Casual. It's no problem for me now, I just hadn't had a chance to try out the upgraded version and the non-`calc` menus before and wanted to do so because they look quite useful. 

FYI the Straight package manager allows me to install directly from the package repository (i.e. this GitHub repo, though I hear that `package.el` now has this functionality too) rather than via MELPA, and I believe that's what the Straight "recipe"

```
(casual :type git :flavor melpa :host github :repo "kickingvegas/casual")
```

indicates. To double check this isn't a MELPA problem, I manually cloned the repo  and got Straight to build and use this local version instead (with head at commit `006b3d4ba15969946c9aa9255c8ac6b1ac82a409`), and unfortunately I get the same problem. Also, when the package is built, the `*Warnings*` buffer pops up with a bunch of messages like:

```
⛔ Warning (comp): ~/.emacs.d/straight/build/casual/casual-editkit.el: Error: Symbol's value as variable is void casual-editkit-navigation-group
⛔ Warning (comp): ~/.emacs.d/straight/build/casual/casual-editkit-utils.el: Error: Symbol's value as variable is void casual-editkit-navigation-group
...
⛔ Warning (comp): ~/.emacs.d/straight/build/casual/casual-calendar.el: Error: Symbol's value as variable is void casual-calendar--navigation-group
⛔ Warning (comp): ~/.emacs.d/straight/build/casual/casual-calendar-utils.el: Error: Symbol's value as variable is void casual-calendar--navigation-group
...
⛔ Warning (comp): ~/.emacs.d/straight/build/casual/casual-calendar.el: Error: Symbol's value as variable is void casual-calendar--navigation-group
⛔ Warning (comp): ~/.emacs.d/straight/build/casual/casual-calendar-utils.el: Error: Symbol's value as variable is void casual-calendar--navigation-group
...
⛔ Warning (comp): ~/.emacs.d/straight/build/casual/casual-calc.el: Error: Symbol's value as variable is void casual-calc-operators-group
⛔ Warning (comp): ~/.emacs.d/straight/build/casual/casual-calc-vector.el: Error: Symbol's value as variable is void casual-calc-operators-group
```

~~which naively don't seem to be related to the dependency on Transient, but I haven't really checked.~~ EDIT: Actually, this is probably related to your switch to `transient-define-group` in the still un-released version of Transient, right? So hopefully everything works in June.

Thanks for all your work on this!

**Reply by deen1** [reply]
Just to briefly add: I checked out the commit associated with version `2.4.2`, before the shift to `transient-define-group` and depending on an older version of Transient, and now the `calc` tmenu works perfectly fine, as do the other ones.

**Reply by kickingvegas** [reply]
@deen1 yes, pulling the older release (2.4.2) should work with Transient < 0.9.0. The breaking change is in using `transient-define-group` (which I approve of).

Package management in Emacs is quite "wild-west"; as most users get Casual direct from MELPA, in real-world practice MELPA is *released* code from a user perspective, no matter what meaning developers wish to impose on it and users. My QA of Casual and its dependencies is based on what is in MELPA and *not* the versioned code in MELPA stable. While this might be considered cardinally "wrong" from a developer lens, it ensures the highest stability for the majority (~95%) of Casual users.
